main/midi/midiPlayer.c

```diff
@@ -0,0 +1,1194 @@
+#include "midiPlayer.h"
+
+#include <string.h>
+#include <inttypes.h>
+
+#include "waveTables.h"
+#include "midiNoteFreqs.h"
+#include "midiUtil.h"
+#include "hdw-dac.h"
+#include "fp_math.h"
+#include "esp_log.h"
+#include "drums.h"
+#include "macros.h"
+
+#define OSC_DITHER
+
+//==============================================================================
+// Generated by tools/midi/freqs.py
+//==============================================================================
+#ifdef OSC_DITHER
+// Apply a random offset to each oscillator to maybe make it less likely for waves to "stack" exactly
+static const uint8_t oscDither[] = {
+    139, 227, 5,   103, 241, 67, 251, 109, 197, 59,  61,  3,   53,  229, 127, 23,  73,  223,
+    13,  19,  47,  7,   181, 37, 2,   239, 29,  113, 167, 131, 41,  151, 83,  137, 11,  193,
+    107, 17,  191, 43,  101, 71, 233, 179, 97,  79,  31,  211, 163, 157, 89,  199, 149, 173,
+};
+#endif
+//==============================================================================
+// End generated code section
+//==============================================================================
+
+// For MIDI values with coarse and fine bytes, each 7 bits
+#define UINT14_MAX (0x3FFF)
+
+/// @brief Convert the sample count to MIDI ticks
+#define SAMPLES_TO_MIDI_TICKS(n, tempo, div) ((n) * 1000000 * (div) / DAC_SAMPLE_RATE_HZ / (tempo))
+
+/// @brief Calculate the number of DAC samples in the given number of milliseconds
+#define MS_TO_TICKS(ms) ((ms) * DAC_SAMPLE_RATE_HZ / 1000)
+
+#define VS_ANY(statePtr) ((statePtr)->on)
+
+// Values for the percussion special states bitmap
+#define SHIFT_HI_HAT   (0)
+#define SHIFT_WHISTLE  (6)
+#define SHIFT_GUIRO    (12)
+#define SHIFT_CUICA    (18)
+#define SHIFT_TRIANGLE (24)
+
+#define MASK_HI_HAT   (0x3F)
+#define MASK_WHISTLE  (0x3F << SHIFT_WHISTLE)
+#define MASK_GUIRO    (0x3F << SHIFT_GUIRO)
```

Comment:
Thought you were making up words, but it's a real instrument.

Reply:
They got all sorts of crazy percussion instruments here!